pkg/nodeagent/controller/healthcheck/kubelet.go

```diff
@@ -0,0 +1,270 @@
+// Copyright 2023 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package healthcheck
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"net/netip"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/gardener/gardener/pkg/nodeagent/dbus"
+)
+
+const (
+	// kubeletServiceName is the systemd service name of the kubelet.
+	kubeletServiceName = "kubelet.service"
+	// defaultKubeletHealthEndpoint is the health endpoint of the kubelet.
+	defaultKubeletHealthEndpoint = "http://127.0.0.1:10248/healthz"
+	// maxToggles defines how often the kubelet can change the readiness during toggleTimeSpan until the node will be rebooted.
+	maxToggles = 5
+	// toggleTimeSpan is a floating time window where the kubelet readiness toggles are considered harmful.
+	toggleTimeSpan = 10 * time.Minute
+)
+
+// KubeletHealthChecker configures the kubelet healthcheck.
+type KubeletHealthChecker struct {
+	// Clock exported for testing.
+	Clock clock.Clock
+	// KubeletReadinessToggles contains an entry for every toggle between kubelet Ready->NotReady or NotReady->Ready state.
+	KubeletReadinessToggles []time.Time
+	// NodeReady indicates if the node is ready. Exported for testing.
+	NodeReady bool
+
+	client     client.Client
+	httpClient *http.Client
+	// firstFailure stores the time of the first failed kubelet health check
+	firstFailure *time.Time
+	dbus         dbus.DBus
+	recorder     record.EventRecorder
+	// lastInternalIP stores the node internalIP
+	lastInternalIP netip.Addr
+	// getAddresses is a func which returns a slice of ip addresses on this node
+	getAddresses          func() ([]net.Addr, error)
+	kubeletHealthEndpoint string
```

Comment:
Nit: Please add doc strings for all fields for the sake of consistency, or remove all doc strings 😄